### PR TITLE
Lab06 - Serial Peripheral Interface

### PR DESCRIPTION
**Pull request by TeodoraMiu**
### Pull Request Overview

This pull request adds lab06 - Serial Peripheral Interface.

### Testing Strategy

This pull request was tested by running lab material examples locally.

### TODO or Help Wanted

This pull request still needs review and suggestions for exercises.

### Build

- [x] Ran `npm build`.

### Author

Signed-off-by: Teodora Miu <teodora.miu@stud.fils.upb.ro>

### CODE REVIEW

File: docs/lab/06/index.md [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+- **CLK** - clock - generated by the main device - subs use it to synchronously read and write to the bus
+- **MOSI** - **M**ain **O**ut **S**ub **I**n - used to carry data from the main device to *all* the subs
+- **MISO** - **M**ain **I**n **S**ub **O**ut - used to carry data from the *active* sub to the main device
+- **CS** - chip select - selects the sub to be activated, or the sub that the main wants to communicate with
+                       - one wire per sub
+                       - inactive subs will disconnect from the *MOSI* and *MISO* lines
+
+:::info
+The communication is *full-duplex*. This means that, during data transmission, both the master and the slave **must write** to the MOSI/MISO wires, no matter if that data is relevant or not.
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:33]:
```suggestion
The communication is *full-duplex*. This means that, during data transmission, both the main and the sub **must write** to the MOSI/MISO wires, no matter if that data is relevant or not.
```

File: docs/lab/06/index.md [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+:::
+
+#### Transmission example
+
+1. main sets `CS` to `LOW` - the sub that the main wants to communicate with is activated - all other subs disconnect from `MOSI` and `MISO` lines
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:38]:
```suggestion
1. main sets `CS` to `LOW` - the sub that the main wants to communicate with is activated - all other subs are already disconnect from `MOSI` and `MISO` lines as their CS line is not active
```

File: docs/lab/06/index.md [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+4. on *rising edge* 
+    - main reads data from the `MISO` line 
+    - sub reads data from the `MOSI` line
+5. on *falling edge*
+    - main writes data to the `MOSI` line
+    - sub writes data to the `MISO` line
+6. repeat steps 4 and 5 until main stops the clock
+
+:::note
+Whether steps 4 and 5 happen on rising/falling edge depends on the configuration of the SPI.
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:50]:
```suggestion
Whether steps 4 and 5 happen on rising/falling edge depends on the `CPOL` configuration of the SPI.
```

File: docs/lab/06/index.md [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+| 3 | 1 | 1 |
+
+| | | |
+|-|-|-| 
+| `CPOL` | Clock polarity | defines when the data bit is read <br/> 0: *rising edge* <br/> 1: *falling edge* |
+| `CPHA` | Clock phase | defines when the data is written to the line <br/> 0: when `CS` *activates* <br/> 1: on *clock edge* (depends on `CPOL`) |
+
+### Daisy Chaining
+
+An SPI network with many sub devices can prove impractical from a hardware point of view. **Daisy chaining** is a method that solves this by linking multiple devices together with the following configuration:
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:75]: This does not solve the problem of many subs, it just provides another way of linking subs. Subs here usually send in MISO the previous MOSI bit.

File: docs/lab/06/index.md [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+
+If main wants to communicate with the second sub, it will send its data through the first sub, and receive data from the second sub through all subs following it. (TODO: explain)
+
+## DMA
+
+When it comes to standard communication protocols that rely on [MMIO](/docs/lab/02), it would be inefficient to rely on the MCU itself to handle all of these data transfers. This is why the **Direct Memory Access** (DMA) is used. Its purpose is to offload the MCU by dealing with simple transmission to and from peripherals, and whenever it finishes a transfer, it raises an interrupt.
+
+## SPI in Embassy
+
+In Embassy, we can use the SPI both blocking and asynchronously. The following example will be using the asynchronous version.
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:89]: Add a warning explaining why the synchronous versions is not good.

File: docs/lab/06/index.md [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+
+:::tip
+The pressure value is split into 3 registers: `press_msb`, `press_lsb` and `press_xlsb`. 
+- `press_msb` is the first binary half of the pressure value (stands for pressure most significant bits)
+- `press_lsb` is the second binary half of the pressure value (stands for pressure least significant bits)
+- `press_xlsb` is an extra degree of precision for the pressure value; **we won't be using it**
+
+To compute the pressure value, we need to read `press_msb` and `press_lsb`, and **add them together**.
+
+*pressure = `press_msb` + `press_lsb`*
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:350]:
```suggestion
*pressure = `press_msb << 8` + `press_lsb`*
```

File: docs/lab/06/index.md [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+- Write the `ctrl_meas` register with appropiate configuration values. You can find information on the contents you should write to this register at section 4.3.4 of the datasheet. (**2p**)
+- Read the pressure value and print it over the serial. (**1p**)
+
+:::tip
+The pressure value is split into 3 registers: `press_msb`, `press_lsb` and `press_xlsb`. 
+- `press_msb` is the first binary half of the pressure value (stands for pressure most significant bits)
+- `press_lsb` is the second binary half of the pressure value (stands for pressure least significant bits)
+- `press_xlsb` is an extra degree of precision for the pressure value; **we won't be using it**
+
+To compute the pressure value, we need to read `press_msb` and `press_lsb`, and **add them together**.
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:348]:
```suggestion
To compute the pressure value, we need to read `press_msb` and `press_lsb`, and **shift and add them together**.
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/lab/06/index.md [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+
+## Serial Peripheral Interface (SPI)
+
+The Serial Peripheral Interface (or SPI) is a communication protocol used between microcontrollers and peripheral ICs (integrated circuits). It allows synchronous transfer of data between one *main* device and one or more *sub* devices (e.g. sensors). The main device is the one that dictates the communication.
+
+### Data transmission
+
+![SPI_protocol](images/spi_network.svg)
+
+During SPI transmission, there are 4 wires used:
```
Comment (DTeodor-Alexaandru) [inline docs/lab/06/index.md:23]: Not necessarily, elaborate about half-duplex, full-duplex and transmission without CS.

File: docs/lab/06/index.md (hunk 1 above)
Comment (DTeodor-Alexaandru) [inline docs/lab/06/index.md:33]: explain why... for every clk pulse dictated by main, 1 bit must be written

**Review by DTeodor-Alexaandru — CHANGES_REQUESTED** [review]

File: docs/lab/06/index.md [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+Whether steps 4 and 5 happen on rising/falling edge depends on the `CPOL` configuration of the SPI.
+:::
+
+![SPI_example](images/spi_network_transmission.svg)
+
+In the image above, the main first exchages data with the first sub by setting its corresponding `CS` line to low. After it finishes transmission with sub 1, it deactivates it and, afterwards, initiates communication with the second sub by setting its `CS` line to low. Notice that both main and sub read and write to the MOSI/MISO line simultaneously.
+
+### SPI modes
+
+SPI has 4 different modes which define when data is read or written. These modes can be configured through 2 registers: `CPOL` and `CPHA`.
```
Comment (DTeodor-Alexaandru) [inline docs/lab/06/index.md:59]: For this section please take a look at https://www.analog.com/en/resources/analog-dialogue/articles/introduction-to-spi-interface.html at Clock Polarity and Clock Phase

File: docs/lab/06/index.md (hunk 5 above)
Comment (TeodoraMiu) [reply docs/lab/06/index.md:97]: Done.

File: docs/lab/06/index.md (hunk 4 above)
Comment (TeodoraMiu) [reply docs/lab/06/index.md]: Removed the part about hardware impracticality and provided a clearer explanation of the way daisy chaining works. Let me know if I should add anything else here.

File: docs/lab/06/index.md (hunk 8 above)
Comment (TeodoraMiu) [reply docs/lab/06/index.md]: I explained that there is also a 3 wire SPI configuration, that provides a half-duplex communication. Anything else I should add?

File: docs/lab/06/index.md (hunk 1 above)
Comment (TeodoraMiu) [reply docs/lab/06/index.md]: Done. Is it clear now?

File: docs/lab/06/index.md (hunk 9 above)
Comment (TeodoraMiu) [reply docs/lab/06/index.md:63]: Fixed it.

**Review by DTeodor-Alexaandru — DISMISSED** [review]

File: docs/lab/06/index.md [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+| `CSB` | `CS` line |
+| `SDO` | `MISO` line |
+
+![bmp280_wiring](images/bmp280_wiring.png)
+
+:::note
+The BMP280 can also be interfaced through I2C, using the same pins but with different functions.
+:::
+
+The Raspberry Pi Pico has two usable SPI channels. Each channel has a set SPI control pins.
```
Comment (alexandruradovici) [inline docs/lab/06/index.md:259]:
```suggestion
The Raspberry Pi Pico has two usable SPI channels. Each channel has a set of SPI control pins.
```

I'm not sure what this wants to mean.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Review by alexandruradovici — APPROVED** [review]

**Review by DTeodor-Alexaandru — APPROVED** [review]

**Pull request merged**